pkgs/by-name/gi/gitbutler/package.nix

```
@@ -0,0 +1,177 @@
+
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+    outputHashes = {
+      "tauri-plugin-context-menu-0.7.1" = "sha256-vKfq20hrFLmfoXO94D8HwAE3UdGcuqVZf3+tOBhLqj0=";
+      "tauri-plugin-log-0.0.0" = "sha256-Mf2/cnKotd751ZcSHfiSLNe2nxBfo4dMBdoCwQhe7yI=";
+    };
+  };
+
+  pnpmDeps = pnpm_9.fetchDeps {
+    inherit pname version src;
+    hash = "sha256-+4SPDvhpvR1bjwTcyLO3HeBuZGL3P03OaEnOvcX9t9c=";
+  };
+
+  nativeBuildInputs = [
+    cargo-tauri
+    desktop-file-utils
+    jq
+    moreutils
+    nodejs
+    pkg-config
+    pnpm_9.configHook
+    wrapGAppsHook3
+  ];
+
+  buildInputs =
+    [ openssl ]
+    ++ lib.optionals stdenv.isLinux [
+      glib-networking
+      libsoup
+      webkitgtk
+    ]
+    ++ lib.optionals stdenv.isDarwin (
+      with darwin.apple_sdk.frameworks;
+      [
+        AppKit
+        CoreServices
+        Security
+        WebKit
+      ]
+    );
+
+  env = {
+    # `pnpm`'s `fetchDeps` and `configHook` uses a specific version of pnpm, not upstream's
+    COREPACK_ENABLE_STRICT = 0;
+
+    # we depend on nightly features
+    RUSTC_BOOTSTRAP = 1;
+
+    # we also need to have `tracing` support in `tokio` for `console-subscriber`
+    RUSTFLAGS = "--cfg tokio_unstable";
+
+    tauriBundle =
+      {
+        Linux = "deb";
+        Darwin = "app";
+      }
+      .${stdenv.hostPlatform.uname.system}
+        or (throw "No tauri bundle available for ${stdenv.hostPlatform.uname.system}");
+
+    ESBUILD_BINARY_PATH = lib.getExe (
+      esbuild.override {
+        buildGoModule =
+          args:
+          buildGoModule (
+            args
+            // rec {
+              version = "0.18.20";
+              src = fetchFromGitHub {
+                owner = "evanw";
+                repo = "esbuild";
+                rev = "v${version}";
+                hash = "sha256-mED3h+mY+4H465m02ewFK/BgA1i/PQ+ksUNxBlgpUoI=";
+              };
+
+              vendorHash = "sha256-+BfxCyg0KkDQpHt/wycy/8CTG6YBA/VJvJFhhzUnSiQ=";
+            }
+          );
+      }
+    );
+  };
+
+  buildPhase = ''
+    runHook preBuild
+
+    cargo tauri build --bundles "$tauriBundle"
+
+    runHook postBuild
+  '';
+
+  installPhase =
+    ''
+      runHook preInstall
+    ''
+    + lib.optionalString stdenv.isDarwin ''
+      mkdir -p $out/bin
+      cp -r target/release/bundle/macos $out/Applications
+      mv $out/Applications/GitButler.app/Contents/MacOS/GitButler $out/bin/git-butler
+      ln -s $out/bin/git-butler $out/Applications/GitButler.app/Contents/MacOS/GitButler
+    ''
+    + lib.optionalString stdenv.isLinux ''
+      cp -r target/release/bundle/"$tauriBundle"/*/data/usr $out
+
+      desktop-file-edit \
+        --set-comment "A Git client for simultaneous branches on top of your existing workflow." \
+        --set-key="Keywords" --set-value="git;" \
+        --set-key="StartupWMClass" --set-value="GitButler" \
+        $out/share/applications/git-butler.desktop
+    ''
+    + ''
+      runHook postInstall
+    '';
+
```

the UI doesn't render correctly on Linux with an Nvidia card. you have to make a wrapper script and disable the WebKit composting mode.

```suggestion

  # WEBKIT_DISABLE_COMPOSITING_MODE essential in NVIDIA + compositor https://github.com/NixOS/nixpkgs/issues/212064#issuecomment-1400202079
  postFixup = ''
    wrapProgram "$out/bin/gitbulter" \
      --set WEBKIT_DISABLE_COMPOSITING_MODE 1
  '';

```

---

i have actually only tested this myself on nixos and nvidia, so i'm pretty surprised to see others having issues with this

unless it's very common, i don't think we should be adding this workaround to each and every tauri and/or webkitgtk app - `modrinth-app` for example does fine without this, along with `cinny`. it'd be much better left to individual configurations where these bugs actually occur

---

If this PR is still open in ~4 days, I'm happy to test on a RTX3060 to confirm.

---

>  unless it's very common

i have GTX 1660 TI with 535.154.05 drivers, and it has happened to me with every tauri app i tried with nix 'till now, including this one.
 
> i don't think we should be adding this workaround to each and every tauri and/or webkitgtk app - `modrinth-app` for example does fine without this, along with `cinny`. it'd be much better left to individual configurations where these bugs actually occur

the problem is that the error messages are pretty (generic), actually, it doesn't even give an error is just spamming the `stdout` with `AcceleratedSurfaceDMABuf was unable to construct a complete framebuffer` and you have to do some search to figure out what this means and how to fix it, I have no problem to do this with other imperative distros, but with Nix, I expect to things to just work so I can use them real fast with `nix run`, `nix shell` to get the job done without all this hassle

---

> i have GTX 1660 TI with 535.154.05 drivers

i think that's your problem: your drivers are outdated. the 535 series is no longer supported, so you should be upgrading to either the feature, production, or beta branch (545, 550, and 555 respectively) of the driver. i have a 1660 super on 550 (which `config.boot.kernelPackages.nvidiaPackages.latest` currently refers to) and this works as expected

> the problem is that the error messages are pretty (generic), actually, it doesn't even give an error is just spamming the stdout with AcceleratedSurfaceDMABuf was unable to construct a complete framebuffer and you have to do some search to figure out what this means and how to fix it, I have no problem to do this with other imperative package managers, but with Nix, I expect to things to just work

i can get this. after all, i've dealt with a lot of bugs like these myself with my card over the years 😆 

but i don't think it should be expected for nixpkgs maintainers to keep up with all of these possible issues and adopt the workarounds (that may or may not affect most users and could give an unexpected experience) into their packages. in this case specifically, it seems like it doesn't affect many people, requires a drastic action of disabling composting entirely, can be fixed with one environment variable, and may only be the result of an outdated driver. i don't think this is a fair tradeoff

i think a better solution for this case (and others) would be to help make users aware and avoid endless searching by making an issue. as 535 is actually still available in nixpkgs (i didn't know this until i started writing this part lol) and this may affect all webkitgtk apps, this should probably have it's own dedicated thread that will be easier to find, provide these workarounds, and notify maintainers so they can look into possible fixes as well

> so I can use them real fast with nix run, nix shell to get the job done without all this hassle

it should work with `nix-shell` if `WEBKIT_DISABLE_COMPOSITING_MODE` is already set in your environment - you can even do this declaratively via `environment.sessionVariables`

with `nix shell` you may need to use `--impure`, though

---

> you should be upgrading to either the feature, production, or beta branch (545, 550, and 555 respectively) of the driver. i have a 1660 super on 550 (which config.boot.kernelPackages.nvidiaPackages.latest currently refers to) and this works as expected

I can't, the latest versions have problems with my card, plus am not sure that this problem has anything to do with the GPU driver version

---

@isabelroses Could you test this?

---

Ahh yeah totally forgot about this PR. I'm happy to.

---

I can confirm without this suggestion it still works perfectly.

---

maybe it is a problem with my particular card